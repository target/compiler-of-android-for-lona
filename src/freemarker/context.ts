export type ContextData = { [key: string]: any }

export class Context {
  #data: ContextData

  constructor(data: ContextData) {
    this.#data = data
  }

  get(key: string): any {
    return this.#data[key]
  }

  has(key: string): boolean {
    return key in this.#data
  }

  set(key: string, value: any) {
    this.#data[key] = value
  }

  withDefaults(data: ContextData): Context {
    return new Context({ 
      ...data,
      ...this.#data
    })
  }
}
