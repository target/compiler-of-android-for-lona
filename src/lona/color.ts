import { ColorValue } from '@lona/compiler/lib/plugins/tokens/tokens-ast'
import * as XML from '../xml/ast'
import {
  createColor,
  cssToHexColor,
  formatQualifiedName,
  Options,
} from '../android/valueResources'

export type Token = {
  qualifiedName: string[]
  value: ColorValue
}

export const convert = (token: Token, options: Options): XML.Element => {
  return createColor(
    formatQualifiedName(token.qualifiedName, options),
    cssToHexColor(token.value.css)
  )
}
