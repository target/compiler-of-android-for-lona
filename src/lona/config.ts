import fs from 'fs'
import path from 'path'
import * as Options from '../options'

export type Config = {
  ignore: string[]
  android: Partial<Options.Raw>
}

export async function getConfig(workspacePath: string): Promise<Config> {
  const configPath = path.join(workspacePath, 'lona.json')

  const data = JSON.parse(await fs.promises.readFile(configPath, 'utf-8'))

  if (!data.ignore) {
    data.ignore = ['**/node_modules/**', '**/.git/**']
  }

  if (!data.android) {
    data.android = {}
  }

  if (data.android.output) {
    if (!data.android.output.startsWith('/')) {
      data.android.output = path.join(workspacePath, data.android.output)
    }
  }

  return data
}
