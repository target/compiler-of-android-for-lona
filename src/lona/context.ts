// import { Config } from "@lona/compiler/lib/utils/config"
// import { Reporter } from "@lona/compiler/lib/helpers/reporter"
// import path from 'path'
// import fs from 'fs'
// import * as LogicAST from "@lona/compiler/lib/helpers/logic-ast"
// import * as LogicScope from "@lona/compiler/lib/helpers/logic-scope"
// import * as LogicUnify from "@lona/compiler/lib/helpers/logic-unify"
// import * as LogicEvaluate from "@lona/compiler/lib/helpers/logic-evaluate"

// export const generate = (config: Config, reporter: Reporter) => {
//   const standardLibsPath = path.join(__dirname, '../../static/logic')
//   const standardLibs = fs.readdirSync(standardLibsPath)

//   const libraryFiles: LogicAST.AST.Program[] = standardLibs.map(
//     x =>
//       LogicAST.makeProgram(
//         JSON.parse(fs.readFileSync(path.join(standardLibsPath, x), 'utf8'))
//       ) as LogicAST.AST.Program
//   )

//   const standardLibsProgram = LogicAST.joinPrograms(libraryFiles)

//   const logicPrograms = Object.keys(config.logicFiles)
//     .map(k => {
//       const node = LogicAST.makeProgram(config.logicFiles[k])
//       if (!node) {
//         return undefined
//       }
//       return {
//         in: k,
//         node,
//       }
//     })
//     .filter(x => !!x)

//   const scopeContext = LogicScope.build(
//     [{ node: standardLibsProgram, in: 'standard library' }].concat(logicPrograms),
//     reporter
//   )

//   const programNode = LogicAST.joinPrograms([
//     standardLibsProgram,
//     ...logicPrograms.map(x => x.node),
//   ])

//   const unificationContext = LogicUnify.makeUnificationContext(
//     programNode,
//     scopeContext,
//     reporter
//   )
//   const substitution = LogicUnify.unify(
//     unificationContext.constraints,
//     reporter
//   )

//   const evaluationContext = LogicEvaluate.evaluate(
//     programNode,
//     programNode,
//     scopeContext,
//     unificationContext,
//     substitution,
//     reporter
//   )

//   return evaluationContext
// }
