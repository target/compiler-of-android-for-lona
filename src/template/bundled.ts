import path from 'path'

export type BundledTemplateNames = 'project' | 'module'

export function templatePathForName(
  templateName: BundledTemplateNames
): string {
  switch (templateName) {
    case 'project':
      return path.join(
        __dirname,
        '../../templates/android-studio/gradle-projects/NewAndroidProject'
      )
    case 'module':
      return path.join(
        __dirname,
        '../../templates/android-studio/gradle-projects/NewAndroidModule'
      )
  }
}
