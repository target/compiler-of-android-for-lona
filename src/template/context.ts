import path from 'path'
import { Context, ContextData } from '../freemarker'
import escape from 'lodash.escape'
import snakeCase from 'lodash.snakecase'

export type CreateTemplateContextOptions = {
  packageName: string
  minSdk: number
  targetSdk: number
  buildSdk: number
  overrides?: ContextData
}

export function createTemplateContext({
  packageName,
  minSdk,
  targetSdk,
  buildSdk,
  overrides,
}: CreateTemplateContextOptions): Context {
  const projectParts = packageName.split('.')
  const projectName = projectParts[projectParts.length - 1]
  const projectPrefix = projectName || '.'

  return new Context({
    topOut: '.',
    projectOut: `./${projectPrefix}`,
    srcDir: `./${projectPrefix}/src`,
    manifestDir: `./${projectPrefix}/src/main`,
    resDir: `./${projectPrefix}/src/main/res`,
    repositoryList: ['google()', 'jcenter()'],
    baseTheme: 'none',
    makeIgnore: true,
    createActivity: false,
    sdkDir: process.env.ANDROID_HOME,
    appTitle: projectName,
    projectName: projectName,
    simpleName: projectName,
    buildApi: buildSdk,
    buildApiString: `${buildSdk}`,
    buildToolsVersion: `${buildSdk}`, // TODO: Should this be 29.0.0?
    packageName: packageName,
    WearprojectName: '',
    minApi: `${minSdk}`,
    minApiLevel: minSdk,
    targetApi: targetSdk,
    targetApiString: `${targetSdk}`,
    gradlePluginVersion: '3.6.0+',
    javaVersion: '1.7',
    enableProGuard: false,
    dependencyList: undefined,
    unitTestsSupported: false,
    assetName: undefined,
    mavenUrl: 'mavenCentral',
    isLibraryProject: true,
    copyIcons: false,
    appCompat: false,
    perModuleRepositories: undefined,
    templateRoot: path.join(__dirname, '../../templates/android-studio'),
    ...overrides,
    // TODO: Implement all helper methods
    escapeXmlAttribute: (value: string) => escape(value),
    escapeXmlString: (value: string) => escape(value),
    escapePropertyValue: (value: string) => value,
    hasDependency: (value: string) => false,
    slashedPackageName: (packageName: string): string =>
      packageName.replace(/\./g, '/'),
    compareVersions: (a: string, b: string): boolean => false,
    activityToLayout: (activityClass: string): string =>
      snakeCase(activityClass),
  })
}
