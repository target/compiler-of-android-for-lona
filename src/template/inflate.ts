import fs from 'fs'
import path from 'path'
import { createFs, IFS, copy } from 'buffs'
import * as XML from '../xml'
import * as FreeMarker from '../freemarker'

import { Config, getConfig } from './config'
import { parse as parseRecipe, Recipe } from './recipe'
import { parse as parseGlobals, Globals } from './globals'

type CreateContextOptions = { packageName: string }

function createFreemarkerContext(
  { packageName }: CreateContextOptions,
  globals: Globals = {}
): FreeMarker.Context {
  const projectParts = packageName.split('.')
  const projectName = projectParts[projectParts.length - 1]
  const projectPrefix = projectName || '.'

  return FreeMarker.createContext({
    ...globals,
    topOut: '.',
    projectOut: `./${projectPrefix}`,
    srcDir: `./${projectPrefix}/src`,
    manifestDir: `./${projectPrefix}/src/main`,
    resDir: `./${projectPrefix}/src/main/res`,
    repositoryList: ['google()', 'jcenter()'],
    baseTheme: 'none',
    makeIgnore: true,
    createActivity: false,
    sdkDir: process.env.ANDROID_HOME,
    appTitle: projectName,
    projectName: projectName,
    simpleName: projectName,
    buildApi: 29,
    buildApiString: '29',
    buildToolsVersion: '29.0.2',
    packageName: packageName,
    WearprojectName: '',
    minApi: '21',
    minApiLevel: 21,
    targetApi: 29,
    targetApiString: '29',
    gradlePluginVersion: '3.6.0+',
    javaVersion: '1.7',
    enableProGuard: false,
    dependencyList: undefined,
    unitTestsSupported: false,
    assetName: undefined,
    mavenUrl: 'mavenCentral',
    isLibraryProject: true,
    copyIcons: false,
    appCompat: false,
    perModuleRepositories: undefined,
    templateRoot: path.join(__dirname, '../../templates/android-studio'),
    escapeXmlAttribute: (value: string) => value,
    escapeXmlString: (value: string) => value,
    escapePropertyValue: (value: string) => value,
    hasDependency: (value: string) => false,
    slashedPackageName: (packageName: string): string =>
      packageName.replace(/\./g, '/'),
    compareVersions: (a: string, b: string): boolean => false,
  })
}

function inflateFMT(
  source: IFS,
  filePath: string,
  context: FreeMarker.Context
): string {
  const data = source.readFileSync(filePath, 'utf8')
  const template = FreeMarker.parse(data)
  if (template.ast.errors) {
    console.log('ERROR: Failed to parse', filePath)
  }
  const inflated = FreeMarker.evaluate(template.ast, context)
  return inflated
}

function readXML(source: IFS, filePath: string): XML.Element {
  const templateString = source.readFileSync(filePath, 'utf8')
  return XML.parse(templateString)
}

function readTemplate(source: IFS, templatePath: string): Config {
  const templateElement = readXML(source, templatePath)
  return getConfig(templateElement)
}

function readRecipe(
  source: IFS,
  recipePath: string,
  context: FreeMarker.Context
): Recipe {
  const inflated = inflateFMT(source, recipePath, context)
  const xml = XML.parse(inflated)
  return parseRecipe(xml)
}

function readGlobals(
  source: IFS,
  globalsPath: string,
  context: FreeMarker.Context
): Globals {
  const inflated = inflateFMT(source, globalsPath, context)
  const xml = XML.parse(inflated)
  return parseGlobals(xml)
}

function execute(
  source: IFS,
  sourcePath: string,
  targetPath: string,
  recipe: Recipe,
  context: FreeMarker.Context
): IFS {
  const { fs: target } = createFs()

  const resolveSourcePath = (filePath: string): string => {
    if (filePath.startsWith('/')) {
      return filePath
    } else {
      return path.join(sourcePath, 'root', filePath)
    }
  }

  const resolveTargetPath = (filePath: string): string => {
    return path.join(targetPath, filePath)
  }

  for (const command of recipe) {
    switch (command.type) {
      case 'dependency':
        const dependencyList = context.get('dependencyList') || []

        if (!dependencyList.includes(command.value)) {
          context.set('dependencyList', [...dependencyList, command.value])
        }

        // console.log('dep', command)
        break
      case 'mkdir':
        // if (!command.value.at) break
        try {
          target.mkdirSync(
            path.join('/', resolveTargetPath(command.value.at)),
            {
              recursive: true,
            }
          )
        } catch {}
        break
      // TODO: Smarter merge
      case 'merge': {
        // console.log('should merge', command.value)

        const mergeSourcePath = resolveSourcePath(command.value.from)
        const mergeTargetPath = resolveTargetPath(command.value.to)

        // console.log('source', mergeSourcePath, 'target', mergeTargetPath)

        const inflated = inflateFMT(source, mergeSourcePath, context)

        const initialText = source.existsSync(mergeTargetPath)
          ? source.readFileSync(mergeTargetPath, 'utf8')
          : ''

        const mergedText = [initialText, inflated]
          .filter(x => x.length > 0)
          .join('\n')

        target.writeFileSync(mergeTargetPath, mergedText)

        break
      }
      case 'instantiate': {
        const inflated = inflateFMT(
          source,
          resolveSourcePath(command.value.from),
          context
        )
        target.mkdirSync(path.dirname(resolveTargetPath(command.value.to)), {
          recursive: true,
        })
        target.writeFileSync(
          resolveTargetPath(command.value.to),
          inflated,
          'utf8'
        )
        break
      }
      case 'copy': {
        // console.log(
        //   'INFO: Copy',
        //   command.value,
        //   resolveSourcePath(command.value.from),
        //   resolveTargetPath(command.value.to)
        // )
        copy(
          fs,
          target,
          resolveSourcePath(command.value.from),
          resolveTargetPath(command.value.to)
        )
        break
      }
    }
  }

  // const inspected = util.inspect(volume.toJSON(), false, null, true)
  // console.log(inspected)

  return target
}

export function inflate(
  source: IFS,
  sourcePath: string,
  targetPath: string,
  options: CreateContextOptions
): IFS {
  const templatePath = path.join(sourcePath, 'template.xml')

  // console.warn('inflating:', templatePath)

  const template = readTemplate(source, templatePath)

  const { globals: globalsName, execute: recipeName } = template

  const globalsPath = path.join(sourcePath, globalsName)

  // console.warn('INFO: reading globals:', globalsPath)

  const globals = readGlobals(
    source,
    globalsPath,
    createFreemarkerContext(options)
  )

  // console.log('INFO: globals', util.inspect(globals, false, null, true))

  const recipePath = path.join(sourcePath, recipeName)

  const sharedContext = createFreemarkerContext(options, globals)

  // console.warn('INFO: reading recipe:', recipePath)

  const recipe = readRecipe(source, recipePath, sharedContext)

  // console.log('recipe', util.inspect(recipe, false, null, true))

  // console.warn('INFO: executing recipe:', recipePath)

  const target = execute(source, sourcePath, targetPath, recipe, sharedContext)

  return target

  // return createFs().fs
}
