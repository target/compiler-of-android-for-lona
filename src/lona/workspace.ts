import { Helpers } from '@lona/compiler/lib/helpers'
import * as Tokens from '@lona/compiler/lib/plugins/tokens'
import { Token } from '@lona/compiler/lib/plugins/tokens/tokens-ast'
import * as FileSearch from '@lona/compiler/lib/utils/file-search'
import { copy, IFS } from 'buffs'
import fs from 'fs'
import path from 'path'
import { Union } from 'unionfs'
import util from 'util'
import {
  addDependenciesAndPluginsToModuleBuildScript,
  addKotlinDependencyToRootBuildScript,
} from '../android/buildScript'
import { formatDrawableName } from '../android/drawableResources'
import { createResourceFiles } from '../android/resources'
import { Validated } from '../options'
import { createFiles as createSvgDrawableFiles } from '../svg/drawable'
import { templatePathForName } from '../template/bundled'
import {
  createTemplateContext,
  CreateTemplateContextOptions,
} from '../template/context'
import { inflate, InflateOptions } from '../template/inflate'
import { createValueResources } from './tokens'

export function inflateProjectTemplate(
  outputPath: string,
  generateGallery: boolean,
  drawableResourceNames: string[],
  options: CreateTemplateContextOptions,
  inflateOptions?: InflateOptions
): { files: IFS; srcPath: string; resPath: string } {
  const { files: projectFiles } = inflate(
    fs,
    templatePathForName('project'),
    outputPath,
    createTemplateContext(options),
    inflateOptions
  )

  const fsAndProject = new Union()
  fsAndProject.use(fs)
  fsAndProject.use(projectFiles)

  const { files: moduleFiles, context: moduleContext } = inflate(
    fsAndProject,
    templatePathForName('module'),
    outputPath,
    createTemplateContext(options),
    inflateOptions
  )

  copy(moduleFiles, projectFiles, '/', '/')

  if (generateGallery) {
    const { files: galleryFiles, context: galleryContext } = inflate(
      fsAndProject,
      templatePathForName('gallery'),
      outputPath,
      createTemplateContext({
        ...options,
        overrides: {
          isNewProject: false,
          layoutName: 'activity_empty',
          activityClass: 'EmptyActivity',
          activityTitle: 'Empty Activity',
          isLauncher: false,
          relativePackage: undefined,
          drawableAssetList: drawableResourceNames,
        },
      }),
      inflateOptions
    )

    addKotlinDependencyToRootBuildScript(
      projectFiles,
      galleryContext,
      outputPath
    )
    addDependenciesAndPluginsToModuleBuildScript(
      projectFiles,
      galleryContext,
      outputPath
    )

    copy(galleryFiles, projectFiles, '/', '/')
  }

  return {
    files: projectFiles,
    srcPath: moduleContext.get('srcDir'),
    resPath: moduleContext.get('resDir'),
  }
}

async function convertTokens(workspacePath: string, helpers: Helpers) {
  const convertedWorkspace = await Tokens.parseWorkspace(
    workspacePath,
    helpers,
    {
      output: false,
    }
  )

  if (!convertedWorkspace) {
    return Promise.reject('Unknown problem while converting tokens')
  }

  const tokens = convertedWorkspace.files.flatMap(file =>
    file.contents.type === 'flatTokens' ? file.contents.value : []
  )

  return tokens
}

async function convertSvgFiles(
  workspacePath: string,
  nameTemplate: string,
  ignore: string[]
): Promise<[string, IFS][]> {
  const svgRelativePaths = FileSearch.sync(workspacePath, '**/*.svg', {
    ignore,
  })

  return Promise.all(
    svgRelativePaths.map(async relativePath => {
      const data = await fs.promises.readFile(
        path.join(workspacePath, relativePath)
      )

      const result: [string, IFS] = [
        relativePath,
        await createSvgDrawableFiles(relativePath, data, { nameTemplate }),
      ]

      return result
    })
  )
}

/**
 * Convert a Lona workspace to an Android library module.
 *
 * @returns A filesystem containing the generated library files.
 */
export async function convert(
  workspacePath: string,
  helpers: Helpers,
  options: Omit<Validated, 'shouldOutputFiles' | 'dryRun'>
): Promise<IFS> {
  const {
    verbose,
    outputPath,
    packageName,
    minSdk,
    buildSdk,
    targetSdk,
    generateGallery,
    valueResourceNameTemplate,
    drawableResourceNameTemplate,
  } = options

  if (verbose) {
    console.log('Converting with options:')
    console.log(util.inspect(options, false, null, true))
  }

  let tokens: Token[]

  try {
    tokens = await convertTokens(workspacePath, helpers)
  } catch (error) {
    helpers.reporter.error('Failed to convert tokens')
    return Promise.reject(error)
  }

  const drawableResources: [string, IFS][] = await convertSvgFiles(
    workspacePath,
    valueResourceNameTemplate,
    helpers.config.ignore
  )

  const drawableResourceNames = drawableResources.map(([key]) =>
    formatDrawableName(key, undefined, {
      nameTemplate: drawableResourceNameTemplate,
    })
  )

  const { files: templateFiles, resPath } = inflateProjectTemplate(
    outputPath,
    generateGallery,
    drawableResourceNames,
    { packageName, minSdk, buildSdk, targetSdk },
    { verbose }
  )

  const valueResources = createValueResources(tokens, {
    minSdk,
    nameTemplate: drawableResourceNameTemplate,
  })

  const { fs: resourceFiles } = createResourceFiles(
    path.join(outputPath, resPath),
    {
      colorResources: valueResources.colors,
      elevationResources: valueResources.elevations,
      textStyleResources: valueResources.textStyles,
      drawableResources,
    }
  )

  copy(resourceFiles, templateFiles, '/', '/')

  return templateFiles
}
