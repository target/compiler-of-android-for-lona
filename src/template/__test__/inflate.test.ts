import fs from 'fs'
import { describe as describeFs } from 'buffs'
import { inflate } from '../inflate'
import { inflateProjectTemplate } from '../../lona/workspace'
import { templatePathForName } from '../bundled'
import { createTemplateContext, CreateTemplateContextOptions } from '../context'

const defaultTemplateOptions: CreateTemplateContextOptions = {
  packageName: 'com.example.designsystem',
  minSdk: 21,
  targetSdk: 29,
  buildSdk: 29,
}

describe('Template / Inflate', () => {
  it('inflates module template', () => {
    const { files } = inflate(
      fs,
      templatePathForName('module'),
      '/prefix',
      createTemplateContext(defaultTemplateOptions)
    )

    expect(describeFs(files, '/')).toMatchSnapshot()
    expect(
      files.readFileSync('/prefix/designsystem/build.gradle', 'utf8')
    ).toMatchSnapshot()
  })

  it('inflates project template', () => {
    const { files } = inflate(
      fs,
      templatePathForName('project'),
      '/prefix',
      createTemplateContext(defaultTemplateOptions)
    )

    expect(describeFs(files, '/')).toMatchSnapshot()
    expect(files.readFileSync('/prefix/build.gradle', 'utf8')).toMatchSnapshot()
  })

  it('inflates project template, then module template on top', () => {
    const { files } = inflateProjectTemplate(
      '/prefix',
      false,
      defaultTemplateOptions
    )

    expect(describeFs(files, '/')).toMatchSnapshot()
    expect(files.readFileSync('/prefix/settings.gradle', 'utf8')).toEqual(
      "include ':designsystem'\n"
    )
  })

  it('inflates project template, then module template, then gallery template', () => {
    const { files } = inflateProjectTemplate(
      '/prefix',
      true,
      defaultTemplateOptions
    )

    const mergedStrings = `<resources>
  <string name="app_name">designsystem</string>
  <string name="title_empty_activity">Empty Activity</string>
  <string name="hello_world">Hello world!</string>
</resources>`

    expect(describeFs(files, '/')).toMatchSnapshot()
    expect(
      files.readFileSync(
        '/prefix/designsystem/src/main/res/values/strings.xml',
        'utf8'
      )
    ).toEqual(mergedStrings)
  })
})
