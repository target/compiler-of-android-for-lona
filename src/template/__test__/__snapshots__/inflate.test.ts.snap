// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Template / Inflate inflates module template 1`] = `
"/ prefix
├── androidTest / com / example / designsystem / ApplicationTest.java
├── designsystem
│   ├── .gitignore
│   ├── build.gradle
│   ├── libs
│   └── src
│       ├── com / example / designsystem
│       └── main
│           ├── AndroidManifest.xml
│           └── res
│               ├── drawable
│               └── values / strings.xml
└── settings.gradle"
`;

exports[`Template / Inflate inflates module template 2`] = `
"
buildscript {
    repositories {
        
        
        google()
        
        jcenter()
        
        

    }
    dependencies {
        classpath 'com.android.tools.build:gradle:3.6.0+'
    }
}


apply plugin: 'com.android.library'



repositories {
        
        
        google()
        
        jcenter()
        
        

}


android {
    compileSdkVersion 29
    buildToolsVersion \\"29.0.2\\"

    defaultConfig {
    
    
        minSdkVersion 21
        targetSdkVersion 29
        versionCode 1
        versionName \\"1.0\\"
    }


}

dependencies {
    
    compile fileTree(dir: 'libs', include: ['*.jar'])


}
"
`;

exports[`Template / Inflate inflates project template 1`] = `
"/ prefix
├── .gitignore
├── build.gradle
├── gradle / wrapper
│   ├── gradle-wrapper.jar
│   └── gradle-wrapper.properties
├── gradle.properties
├── gradlew
├── gradlew.bat
├── local.properties
└── settings.gradle"
`;

exports[`Template / Inflate inflates project template 2`] = `
"// Top-level build file where you can add configuration options common to all sub-projects/modules.

buildscript {
    repositories {
        
        
        google()
        
        jcenter()
        
        

    }
    dependencies {
        classpath 'com.android.tools.build:gradle:3.6.0+'

        // NOTE: Do not place your application dependencies here; they belong
        // in the individual module build.gradle files
    }
}

allprojects {
    repositories {
        
        
        google()
        
        jcenter()
        
        

    }
}
"
`;

exports[`Template / Inflate inflates project template, then module template on top 1`] = `
"/ prefix
├── .gitignore
├── androidTest / com / example / designsystem / ApplicationTest.java
├── build.gradle
├── designsystem
│   ├── .gitignore
│   ├── build.gradle
│   ├── libs
│   └── src
│       ├── com / example / designsystem
│       └── main
│           ├── AndroidManifest.xml
│           └── res
│               ├── drawable
│               └── values / strings.xml
├── gradle / wrapper
│   ├── gradle-wrapper.jar
│   └── gradle-wrapper.properties
├── gradle.properties
├── gradlew
├── gradlew.bat
├── local.properties
└── settings.gradle"
`;
